### ISSUE TITLE
rabbitmq + prefork is sluggish (>=4.0.0)

### ISSUE THREAD

**Issue by diresi** [celery/celery#3737]
## Checklist

- [x] I have included the output of ``celery -A proj report`` in the issue.
      (if you are not able to do this, then at least specify the Celery
       version affected).
- [x] I have verified that the issue exists against the `master` branch of Celery.

## Steps to reproduce
I've prepared a simple test [here](https://github.com/diresi/test-celery4).
```
$ celery -A task report

software -> celery:4.0.2 (latentcall) kombu:4.0.2 py:3.5.2
            billiard:3.5.0.2 py-amqp:2.1.4
platform -> system:Linux arch:64bit, ELF imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:amqp results:rpc:///

result_backend: 'rpc:///'
broker_url: 'amqp://guest:********@localhost:5672//'
```
I'm running on archlinux on a recent kernel:
```
$ uname -a
Linux wir 4.8.13-1-ARCH #1 SMP PREEMPT Fri Dec 9 07:24:34 CET 2016 x86_64 GNU/Linux
```

Start the celery worker with a single `prefork` pool process and run the (only) provided task. The included `test.py` does that and just prints out the timestamps.

The reported timestamps and the worker log show variable but rather long delays between the reception of a task in the worker process and the actual execution of the task in a pool process.
```
[2017-01-04 10:27:06,462: INFO/MainProcess] Received task: task.echo[50829dd2-9e4d-4ae9-993a-2a31587be2af]  
[2017-01-04 10:27:06,928: INFO/PoolWorker-1] Task task.echo[50829dd2-9e4d-4ae9-993a-2a31587be2af] succeeded in 0.025460437999981878s: datetime.datetime(2017, 1, 4, 10, 27, 6, 903513)
...
[2017-01-04 10:27:09,664: INFO/MainProcess] Received task: task.echo[3e0e02af-fa74-466b-ac1f-d8eb45f33a00]  
[2017-01-04 10:27:10,923: INFO/PoolWorker-1] Task task.echo[3e0e02af-fa74-466b-ac1f-d8eb45f33a00] succeeded in 0.0006908990000056292s: datetime.datetime(2017, 1, 4, 10, 27, 10, 922486)
```

Note how the second invocation takes >1s from reception in the worker to actual execution.

This delay is somewhat random, as the loop in `kombu.hub` reacts on other events, but you'll note the delay after a couple of invocations.

To me it seems that the poller (epoll on linux) misses some file descriptor, but that is just a wild guess :).

## Expected behavior

The worker should forward tasks to the pool process(es) immediately when received, not after some random delay. This actually works in celery 3.1.25 where the delay (reception to execution) is around 10ms on my system.

```
$ celery -A task report

software -> celery:3.1.25 (Cipater) kombu:3.0.37 py:3.5.2
            billiard:3.3.0.23 py-amqp:1.4.9
platform -> system:Linux arch:64bit, ELF imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:amqp results:rpc:///

CELERY_RESULT_BACKEND: 'rpc:///'
BROKER_URL: 'amqp://guest:********@localhost:5672//'
```

```
[2017-01-04 10:39:38,685: INFO/MainProcess] Received task: task.echo[549943c9-c6a8-41cd-81a7-79207b634a84]
[2017-01-04 10:39:38,699: INFO/MainProcess] Task task.echo[549943c9-c6a8-41cd-81a7-79207b634a84] succeeded in 0.012934925000081421s: datetime.datetime(2017, 1, 4, 10, 39, 38, 686500)
...
[2017-01-04 10:39:42,040: INFO/MainProcess] Received task: task.echo[493ab87e-8c0c-4743-bfce-dd64768b7aca]
[2017-01-04 10:39:42,041: INFO/MainProcess] Task task.echo[493ab87e-8c0c-4743-bfce-dd64768b7aca] succeeded in 0.0006751369999165036s: datetime.datetime(2017, 1, 4, 10, 39, 42, 40724)
```

## Actual behavior

With celery>=4 random delays occor (reception to execution) when using a prefork pool. I haven't found any option that somehow influences this behavior (-Ofcfs, -E, --without-(gossip|mingle|heartbeat)).

I've tested versions 4.0.2, 4.0.1, 4.0.0 and the current master (9d2566e9c) where delays up to 2s occur (I made sure my system is close-to-idle during my tests).

Note that increasing the request rate (running the task in multiple loops) actually improves this behavior, I can only guess that the async loop receives more events in that case.

**Comment by jalessio** [comment]
Possibly related to #3814?

**Comment by diresi** [comment]
@jalessio Yes, seems like the same issue, although I'm not sure whether #3814 is also using a prefork worker pool.

**Comment by matteius** [comment]
We are also seeing this ... not sure if its prefork related or just celery 4.  We may have to downgrade if there is no suggested work around as it was noticed that the tasks are taking much longer now.

**Comment by matteius** [comment]
Is this definitely tied to prefork, or does the Solo pool also have this issue?

**Comment by matteius** [comment]
I simulated this test and was able to show that message consumption also slows down with the solo Pool, which is slightly slower than the prefork Pool for a single process.   I don't think this bug is confined to Prefork pool ...

**Comment by samson-wang** [comment]
Same problem observed. Random execution delay after Task received. Often get delay > 1s.
The librabbitmq resolved it.

**Comment by edmorley** [comment]
@samson-wang, a new py-amqp release including celery/py-amqp#144 isn't yet out, but does installing from py-amqp master fix the issue for you too?

**Comment by samson-wang** [comment]
@edmorley I didn't try master install. I have tried the solution from https://github.com/celery/py-amqp/pull/144 . It didn't work. There are so many diffs between the v2.1.4 and the master branch. So my try can't be sufficient.

**Comment by 23doors** [comment]
https://github.com/celery/py-amqp/pull/144 is a fix for the issue with 0-2s task delays for task protocol=2 when using pyamqp. If you observe higher delays >2s then this may be something completely separate.

**Comment by diresi** [comment]
I'm actually getting an exception when starting celery:
```
 File "..../venvs/celery-test/lib/python3.6/site-packages/amqp-2.1.4-py3.6.egg/amqp/connection.py", line 349, in _on_start
    self.mechanisms = mechanisms.split(b' ')
TypeError: must be str or None, not bytes
```

This happens with py-amqp@dcd2b56df030 and the test from the mentioned repo, packages installed are:
```
$ pip freeze
amqp==2.1.4
billiard==3.5.0.2
celery==4.0.2
kombu==4.0.2
pytz==2017.2
vine==1.1.3

```

**Comment by 23doors** [comment]
@diresi I am unfamiliar with this part of code but it seems that this line was changed in  https://github.com/celery/py-amqp/pull/110 . Maybe it introduced some python3 incompatibility with some cases. It should probably work the same with:
```
self.mechanisms = mechanisms.split()
```
or maybe it should check for explicit type and convert. Which is safer considering it may depend on mechanisms being bytes type in another part.

@alexsdutton @auvipy any thoughts why this could happen?

**Comment by auvipy** [comment]
I would say to try the masters branch and then let us know whats going on

**Comment by 23doors** [comment]
Yeah, well. Version mentioned above is the latest master.

**Comment by diresi** [comment]
Yes, it was the master branch (from 2 days ago) and it totally seems like a python3 issue. I just fixed this one line but the overall problem with sluggishness doesn't seem to be fixed.
Am I the only one to see this issue with the testcase from the initial report?

**Comment by murdav** [comment]
Same issue for me. 

kombu==4.0.2
billiard==3.5.0.2
celery==4.0.2
But using **amqp==2.1.4 works**, while using amqp==2.2.0 I got the error:

```
[2017-07-13 14:25:04,050: CRITICAL/MainProcess] Unrecoverable error: TypeError('must be str or None, not bytes',)
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/celery/worker/worker.py", line 203, in start
    self.blueprint.start(self)
  File "/usr/local/lib/python3.6/site-packages/celery/bootsteps.py", line 119, in start
    step.start(parent)
  File "/usr/local/lib/python3.6/site-packages/celery/bootsteps.py", line 370, in start
    return self.obj.start()
  File "/usr/local/lib/python3.6/site-packages/celery/worker/consumer/consumer.py", line 318, in start
    blueprint.start(self)
  File "/usr/local/lib/python3.6/site-packages/celery/bootsteps.py", line 119, in start
    step.start(parent)
  File "/usr/local/lib/python3.6/site-packages/celery/worker/consumer/connection.py", line 21, in start
    c.connection = c.connect()
  File "/usr/local/lib/python3.6/site-packages/celery/worker/consumer/consumer.py", line 405, in connect
    conn = self.connection_for_read(heartbeat=self.amqheartbeat)
  File "/usr/local/lib/python3.6/site-packages/celery/worker/consumer/consumer.py", line 412, in connection_for_read
    self.app.connection_for_read(heartbeat=heartbeat))
  File "/usr/local/lib/python3.6/site-packages/celery/worker/consumer/consumer.py", line 436, in ensure_connected
    callback=maybe_shutdown,
  File "/usr/local/lib/python3.6/site-packages/kombu/connection.py", line 405, in ensure_connection
    callback)
  File "/usr/local/lib/python3.6/site-packages/kombu/utils/functional.py", line 333, in retry_over_time
    return fun(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/kombu/connection.py", line 261, in connect
    return self.connection
  File "/usr/local/lib/python3.6/site-packages/kombu/connection.py", line 802, in connection
    self._connection = self._establish_connection()
  File "/usr/local/lib/python3.6/site-packages/kombu/connection.py", line 757, in _establish_connection
    conn = self.transport.establish_connection()
  File "/usr/local/lib/python3.6/site-packages/kombu/transport/pyamqp.py", line 130, in establish_connection
    conn.connect()
  File "/usr/local/lib/python3.6/site-packages/amqp/connection.py", line 302, in connect
    self.drain_events(timeout=self.connect_timeout)
  File "/usr/local/lib/python3.6/site-packages/amqp/connection.py", line 483, in drain_events
    while not self.blocking_read(timeout):
  File "/usr/local/lib/python3.6/site-packages/amqp/connection.py", line 489, in blocking_read
    return self.on_inbound_frame(frame)
  File "/usr/local/lib/python3.6/site-packages/amqp/method_framing.py", line 69, in on_frame
    callback(channel, method_sig, buf, None)
  File "/usr/local/lib/python3.6/site-packages/amqp/connection.py", line 493, in on_inbound_method
    method_sig, payload, content,
  File "/usr/local/lib/python3.6/site-packages/amqp/abstract_channel.py", line 142, in dispatch_method
    listener(*args)
  File "/usr/local/lib/python3.6/site-packages/amqp/connection.py", line 347, in _on_start
    self.mechanisms = mechanisms.split(b' ')
TypeError: must be str or None, not bytes

```

while starting the workers, e.g.:

`exec celery worker -A proj --autoscale=8,0 -Ofair -E -l INFO -n worker.high -Q high --pidfile="/backend/run/celery/high.pid" --logfile="/backend/logs/celery/high.log"
`

**Comment by thedrow** [comment]
@murdav Please upgrade to 2.2.1 which contains a fix for this problem.

**Comment by thedrow** [comment]
Any feedback

**Comment by murdav** [comment]
Sorry I forgot about this! Anyway using amqp==2.2.2 the issue has been resolved.

**Comment by thedrow** [comment]
@diresi Is it the same for you?

**Comment by diresi** [comment]
@thedrow thanks for reminding me. I can confirm that the issue is resolved with amqp==2.2.2 (celery==4.1.0).

thanks!
